### PR TITLE
feat: require content-type parser to set content-type

### PR DESCRIPTION
**Pull request by SgtPooki**
## Title

feat: require content-type parser to set content-type

## Description

* adds `contentTypeParser` function to createVerifiedFetch options & implements it.
* renamed `getStreamAndContentType` to `getStreamFromAsyncIterable` that now returns a stream with the firstChunk seen, so we can pass it to the `contentTypeParser` function.
* updates tests in packages/verified-fetch & packages/interop
* updates packageDocumentation with example

Related https://github.com/ipfs/helia/pull/416
Fixes https://github.com/ipfs/helia/issues/422

## Notes & open questions

- feat: require content-type parser to set content-type
- test: remove vFetch interop content-type assertions
- docs: add content-type customization example

## Change checklist

- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation if necessary (this includes comments as well)
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

**Review by SgtPooki — COMMENTED** [review]
self review

File: packages/verified-fetch/src/index.ts [hunk 1]
```diff
@@ -262,8 +282,16 @@ export interface VerifiedFetch {
 export interface CreateVerifiedFetchWithOptions {
   gateways: string[]
   routers?: string[]
+  /**
+   * A function to handle parsing content type from bytes. The function you provide will be passed the first set of
+   * bytes we receive from the network, and should return a string that will be used as the value for the `Content-Type`
+   * header in the response.
+   */
+  contentTypeParser?: ContentTypeParser
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/verified-fetch/src/index.ts:290]: Perhaps not in the first instance to keep it simple?  We can always add it later.

File: packages/interop/src/verified-fetch-unixfs-dir.spec.ts [hunk 2]
```diff
@@ -1,17 +1,20 @@
...
+import { filetypemime } from 'magic-bytes.js'
 import { createKuboNode } from './fixtures/create-kubo.js'
 import { loadFixtureDataCar } from './fixtures/load-fixture-data.js'
+import type { VerifiedFetch } from '@helia/verified-fetch'
 import type { Controller } from 'ipfsd-ctl'
 
 describe('@helia/verified-fetch - unixfs directory', () => {
   let controller: Controller
-  let verifiedFetch: Awaited<ReturnType<typeof createVerifiedFetch>>
+  let verifiedFetch: VerifiedFetch
```
Comment (achingbrain) [inline packages/interop/src/verified-fetch-unixfs-dir.spec.ts:12]: This type is a bit simpler.

File: packages/verified-fetch/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/verified-fetch/src/index.ts:298]: I'm not sure, I think maybe we want to create a separation between the `createVerifiedFetch` options arg type and the `VerifiedFetch` class init arg type.

File: packages/verified-fetch/src/index.ts [hunk 3]
```diff
@@ -112,6 +112,31 @@
...
+ * @example Customizing content-type parsing
+ *
+ * ```typescript
+ * import { createVerifiedFetch } from '@helia/verified-fetch'
+ * import { fileTypeFromBuffer } from '@sgtpooki/file-type'
+ *
+ * const fetch = await createVerifiedFetch({
+ *  gateways: ['https://trustless-gateway.link'],
+ *  routers: ['http://delegated-ipfs.dev'],
+ *  contentTypeParser: async (bytes) => {
```
Comment (2color) [inline packages/verified-fetch/src/index.ts:132]: Maybe annotate `bytes` here with the type?

**Review by achingbrain — APPROVED** [review]
LGTM.

I made a few changes, please comment if you disagree with them:

1. Split the content type parser tests into their own file
2. Defaulted to a `Content-Type` of `application/octet-stream` instead of not setting one - I think this would be more consistent with what a web server would do
3. Added tests that show how to use `@sgtpooki/file-type` and also `magic-bytes.js` as they both work in browsers
4. Allows the contentTypeParser to be sync as well as async, and to also return `undefined` in which case we fall back to `application/octet-stream`

Also fixed a few bugs:

1. The `contentTypeParser` option wasn't being passed through if you used `createVerifiedFetch`
2. The interop test had guessing the type of an image as `image/png` removed - the file is actually `image/jpeg` with a `.png` extension :trollface:, I guess this was throwing the previous content-type guesser off - now the test tests for the correct mime type

**Comment by achingbrain** [comment]
We may wish to also pass the file name to `contentTypeParser` if it's available?  This way if it's passed `index.html` it can just use the extension to return `text/html` instead of nothing.

File: packages/verified-fetch/src/index.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (SgtPooki) [reply packages/verified-fetch/src/index.ts:298]: thats why I went with this initially. we could always update if we really need to, but i don't imagine createVerifiedFetch init args will vary much

File: packages/verified-fetch/src/index.ts (hunk 3 above)
Comment (SgtPooki) [reply packages/verified-fetch/src/index.ts:132]: it's Uint8Array's all the way down, and the interface has it defined

**Review by SgtPooki — COMMENTED** [review]
i'm good with the changes

File: packages/verified-fetch/src/index.ts (hunk 3 above)
Comment (SgtPooki) [reply packages/verified-fetch/src/index.ts:132]: so its the same for browsers and nodejs

File: packages/interop/src/verified-fetch-unixfs-dir.spec.ts [hunk 4]
```diff
@@ -42,19 +45,37 @@ describe('@helia/verified-fetch - unixfs directory', () => {
...
+        routers: process.env.RUNNER_ENV === 'node' ? [`http://${controller.api.gatewayHost}:${controller.api.gatewayPort}`] : [],
+        contentTypeParser: (bytes) => {
+          return filetypemime(bytes)?.[0]
+        }
+      })
+    })
+
+    it('can return an image content-type for unixfs pathed data', async () => {
+      const resp = await verifiedFetch('ipfs://QmbQDovX7wRe9ek7u6QXe9zgCXkTzoUSsTFJEkrYV1HrVR/1 - Barrel - Part 1.png')
+      // tediously this is actually a jpeg file with a .png extension
```
[the remark this answers is not in the document]
Comment (achingbrain) [reply packages/interop/src/verified-fetch-unixfs-dir.spec.ts:74]: inorite

**Pull request merged**